caas/kubernetes/provider/k8s.go

```diff
@@ -1348,3 +1374,49 @@ func appSecretName(appName, containerName string) string {
 	// A pod may have multiple containers with different images and thus different secrets
 	return "juju-" + appName + "-" + containerName + "-secret"
 }
+
+func mergeConstraintToResources(device devices.KubernetesDeviceParams, resources *core.ResourceRequirements) error {
+	if resources.Limits == nil {
+		resources.Limits = core.ResourceList{}
+	}
+	if resources.Requests == nil {
+		resources.Requests = core.ResourceList{}
+	}
+
+	resourceName := core.ResourceName(device.Type)
+	if v, ok := resources.Limits[resourceName]; ok {
+		logger.Debugf("resource %q - max count %#v has been overwritten by %#v", resourceName, v, device)
+	}
+	if v, ok := resources.Requests[resourceName]; ok {
+		logger.Debugf("resource %q - min count %#v has been overwritten by %#v", resourceName, v, device)
+	}
+	// GPU request/limit have to be set to same value equals to the Count.
+	// - https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/#clusters-containing-different-types-of-nvidia-gpus
+	resources.Limits[resourceName] = *resource.NewQuantity(device.Count, resource.DecimalSI)
+	resources.Requests[resourceName] = *resource.NewQuantity(device.Count, resource.DecimalSI)
+	return nil
+}
+
+func buildNodeSelector(nodeLabel string) map[string]string {
+	// TODO(ycliuhw): to support GKE, set it to `cloud.google.com/gke-accelerator`,
```

Comment:
Does this not work at all on GKE unless we use "cloud.google.com/gke-accelerator"?
Make it TODO(caas) so we can easily find it.
Is it only GKE that's different? How about AWS or our k8s deployment on top of a cloud using our bundles?

Reply:
from https://cloud.google.com/kubernetes-engine/docs/how-to/gpus here, we can see GKE has its own `Key`(GKE kubeconfig fie also has very different format), but Official NVIDIA GPU device plugin uses `accelerator`. 
`nodeSelector` is only required for clusters with multi GPU instance groups configured. 
1. I think AKS only supports single GPU type per cluster because I didn't find anything related from Azure document.
2. K8s clusters running via Kops on AWS is using `accelerator`.
3. EKS is not GA yet, i think.